src/docker.rs

```diff
@@ -0,0 +1,345 @@
+/*
+    Copyright © 2023, ParallelChain Lab
+    Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0
+*/
+
+//! Implements the operations with Docker Containers for building contract, for example, starting container,
+//! copying files to container and executing commands inside docker.
+
+use std::{
+    io::{Read, Write},
+    ops::Not,
+    path::{Path, PathBuf}
+};
+
+use bollard::{
+    container::{
+        Config, CreateContainerOptions, DownloadFromContainerOptions, RemoveContainerOptions,
+        StartContainerOptions, UploadToContainerOptions,
+    },
+    exec::{CreateExecOptions, StartExecOptions},
+    image::CreateImageOptions,
+    service::HostConfig,
+    Docker,
+};
+use futures_util::TryStreamExt;
+use tar::Archive;
+
+use flate2::write::GzEncoder;
+use flate2::Compression;
+use std::fs::File;
+
+use crate::error::Error;
+use rand::{distributions::Alphanumeric, thread_rng, Rng};
+
+const PCHAIN_COMPILE_IMAGE: &str = "parallelchainlab/pchain_compile:mainnet01";
+
+/// get a random string for naming the docker container.
```

Comment:
There are some typos in comments.

line 37: change to "// Generate a random Docker container name"
line 156: change to " // Parse compressed file"
line 165: change to "// Save to destination"